### PR TITLE
Disable custom Git Hooks globally via configuration file

### LINKED ISSUE #2449: Disable custom Git Hooks globally via configuration value
- Gitea version (or commit ref): 04ec795
- Git version: 2.11.0
- Operating system: macOS 10.12.6
- Database (use `[x]`):
  - [x] PostgreSQL
  - [x] MySQL
  - [x] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant

## Description
As a sysadmin hosting gitea for others I'd like the option to disable custom githooks via configuration value so that even the admin of the site can't enable them for users.

### PR DESCRIPTION
**Pull request by techknowlogick**
Fixes #2449 by allowing the disabling of custom git hooks via configuration file

### CODE REVIEW

File: modules/setting/setting.go [hunk 1]
```diff
@@ -817,6 +818,7 @@ func NewContext() {
 	ReverseProxyAuthUser = sec.Key("REVERSE_PROXY_AUTHENTICATION_USER").MustString("X-WEBAUTH-USER")
 	MinPasswordLength = sec.Key("MIN_PASSWORD_LENGTH").MustInt(6)
 	ImportLocalPaths = sec.Key("IMPORT_LOCAL_PATHS").MustBool(false)
+	DisableGitHooks = sec.Key("DISABLE_GIT_HOOKS").MustBool(false)
```
Comment (jonasfranz) [inline modules/setting/setting.go:821]: Please add this also to the `app.ini` file.

File: models/user.go [hunk 2]
```diff
@@ -237,7 +237,7 @@ func (u *User) CanCreateOrganization() bool {
 
 // CanEditGitHook returns true if user can edit Git hooks.
 func (u *User) CanEditGitHook() bool {
-	return u.IsAdmin || u.AllowGitHook
+	return !setting.DisableGitHooks && (u.IsAdmin || u.AllowGitHook)
```
Comment (JonasFranzDEV) [inline models/user.go]: Please add this check to the API to if needed. 
https://github.com/go-gitea/gitea/blob/04ec79579cdcd5cf647ef3ff33d35ecfa7550b49/routers/api/v1/api.go#L377

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: modules/setting/setting.go (hunk 1 above)
Comment (techknowlogick) [reply modules/setting/setting.go:821]: @JonasFranzDEV Thanks for this feedback. I've just pushed a new commit with the option in `app.ini`

File: models/user.go (hunk 2 above)
Comment (techknowlogick) [inline models/user.go:240]: Thanks for this feedback, especially linking me to where I could look. Investigating this (by calling the API to see what the response would be), I found that it only returns the standard gitea/slack/discord/etc.. hooks, and couldn't find it returning the githooks.

**Review by jonasfranz — APPROVED** [review]

**Comment by sapk** [comment]
LGTM

**Review by sapk — APPROVED** [review]

File: conf/app.ini [hunk 3]
```diff
@@ -206,6 +206,8 @@ REVERSE_PROXY_AUTHENTICATION_USER = X-WEBAUTH-USER
 MIN_PASSWORD_LENGTH = 6
 ; True when users are allowed to import local server paths
 IMPORT_LOCAL_PATHS = false
+; Disable all (including admin) users to create custom git-hooks
```
Comment (ethantkoenig) [inline conf/app.ini]: I would suggest `Prevent all users (including admin) from creating custom git hooks`

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
One nit

**Comment by techknowlogick** [comment]
Thanks @ethantkoenig. I've update the comment in app.ini to align with what you suggested. It is now much more clear.

File: templates/admin/user/edit.tmpl [hunk 4]
```diff
@@ -86,7 +86,7 @@
 				<div class="inline field">
 					<div class="ui checkbox">
 						<label><strong>{{.i18n.Tr "admin.users.allow_git_hook"}}</strong></label>
-						<input name="allow_git_hook" type="checkbox" {{if .User.CanEditGitHook}}checked{{end}}>
+						<input name="allow_git_hook" type="checkbox" {{if .User.CanEditGitHook}}checked{{end}} {{if DisableGitHooks}}disabled{{end}}>
```
Comment (bkcsoft) [inline templates/admin/user/edit.tmpl:89]: Since `User.CanEditGitHook` already checks this, is this change needed? 

https://github.com/go-gitea/gitea/pull/2450/files#diff-46259196476f860fea33754fcb22e9eeR240

**Review by bkcsoft — COMMENTED** [review]
Otherwise LGTM

File: templates/admin/user/edit.tmpl (hunk 4 above)
Comment (techknowlogick) [reply templates/admin/user/edit.tmpl:89]: Thanks for the review @bkcsoft. This change is needed so that an admin doesn't get confused. If the global setting is set to disable githooks, then the template change will prevent the admin from checking the box in the admin dashboard and wondering why the change doesn't take effect.

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

File: templates/admin/user/edit.tmpl (hunk 4 above)
Comment (bkcsoft) [reply templates/admin/user/edit.tmpl:89]: aah right, missed the "disabled" part :)

**Comment by bkcsoft** [comment]
@techknowlogick try rebasing on master and see if the build-failures still persist

**Comment by techknowlogick** [comment]
@bkcsoft I've just rebased against master, and the build failed. However, I think that could just be the upgrade for Drone, as I looked at a couple other builds from others and they failed in the same way.

**Comment by techknowlogick** [comment]
@bkcsoft I've just rebased again with the most recent changes and drone is now reporting a success.

**Comment by lafriks** [comment]
@techknowlogick please rabase again

**Comment by techknowlogick** [comment]
@lafriks I've just rebased. The build was cancelled. Should I push again so Drone tries to re-run, or was it cancelled for a reason?

**Comment by lafriks** [comment]
@techknowlogick yes, do force push

**Comment by techknowlogick** [comment]
@lafriks I had force-pushed after this most recent rebase, however as now it is the same as my local I can't force push again. Is there another way to trigger drone?

**Comment by lafriks** [comment]
@techknowlogick you should still be able to do `git push -f origin`

**Comment by techknowlogick** [comment]
@lafriks nevermind. I was able to force push again.

**Comment by lafriks** [comment]
@techknowlogick I'm sorry but you will have to rebase again as other PR build was first :)

**Comment by lafriks** [comment]
Make LG-TM work
